src/next.ts

```diff
@@ -0,0 +1,64 @@
+import areInputsEqual from './are-inputs-equal';
+
+// Using ReadonlyArray<T> rather than readonly T as it works with TS v3
+export type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;
+
+type Cache<TResult> = {
+  lastThis: unknown;
+  lastArgs: unknown[];
+  lastResult: TResult;
+};
+
+type MemoizedFn<T extends (this: any, ...args: any[]) => any> = {
+  name: string;
+  clear: () => void;
+  (...args: Parameters<T>): ReturnType<T>;
+};
+
+function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
+  resultFn: TFunc,
+  isEqual: EqualityFn = areInputsEqual,
+): MemoizedFn<TFunc> {
+  let map = new WeakMap<Record<string, any>, Cache<ReturnType<TFunc>>>();
```

Comment:
I need to test this; but I don't think my usage of weakmap is correct. I am not sure if I can use weakmap for this problem. What do you think @theKashey?

Reply:
it will not do anything here, especially if you never change `key`

However 😉 what will happen if you use `this` as a key? That will allow creating non-overlapping and not-self-wiping caches without much effort.

Reply:
`this` can be undefined, and the key to a weakmap needs to be an `object`

Reply:
I moved to a `Map` for now, but super open to WeakMap

Reply:
In case of not being set(undefined) `this` can default to a locally defined constant key.

Reply:
Interesting